pytorch_pfn_extras/onnx/as_output.py

```diff
@@ -0,0 +1,107 @@
+import onnx
+from typing import List, NamedTuple, Optional
+import torch
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class _Output(NamedTuple):
+    name: str
+    value: torch.Tensor
+
+
+class _Outputs:
+    _values: List[_Output]
+
+    def __init__(self) -> None:
+        self._values = []
+
+    @property
+    def values(self) -> List[_Output]:
+        return self._values
+
+    def add(self, name: str, value: torch.Tensor) -> None:
+        self._values.append(_Output(name, value))
+
+    def add_outputs_to_model(self, onnx_graph: onnx.ModelProto) -> None:
+        if len(self.values) == 0:
+            return
+
+        old_name_to_new_name = {}
+        n_output = len(onnx_graph.graph.output)
+        assert n_output >= len(self.values)
+
+        # Rename last len(self.values) outputs
+        for i, additional_output in enumerate(onnx_graph.graph.output[-len(self.values):]):
+            name = self.values[i].name
+            orig_name = additional_output.name
+            old_name_to_new_name[orig_name] = name
+            additional_output.name = name
+
+        # Rename names in graph
+        for node in onnx_graph.graph.node:
+            for i, v in enumerate(node.input):
+                if v in old_name_to_new_name:
+                    node.input[i] = old_name_to_new_name[v]
+            for i, v in enumerate(node.output):
+                if v in old_name_to_new_name:
+                    node.output[i] = old_name_to_new_name[v]
+
+        for v in onnx_graph.graph.input:
+            if v.name in old_name_to_new_name:
+                v.name = old_name_to_new_name[v.name]
+
+
+_outputs: Optional[_Outputs] = None
```

Comment:
Instead of having this global variable, would it be possible to turn `_start_trace` and `_stop_trace` into a context manager?

Reply:
We can use context manager instead of two separate functions (and I will do refactoring later). But we cannot remove the global variable because it's used in `as_output` function. `as_output` function will be invoked in the user code.

Reply:
If we have the context manager saving the `_outputs` in thread local storage, I think it is easier for the `as_output` function to retrieve it without having a global variable.
What do you think?

Reply:
I see. I think you are right, and I'll fix the implementation.